packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -3504,6 +3084,282 @@ merge_matrices(CrsMatrixStruct<Scalar, LocalOrdinal, GlobalOrdinal, Node>& Aview
+              AddKern::execution_space, AddKern::memory_space, AddKern::memory_space> KKH;
+  KKH handle;
+  handle.create_spadd_handle(false);
+  auto addHandle = handle.get_spadd_handle();
+#ifdef HAVE_TPETRA_MMM_TIMINGS
+  auto MM = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("TpetraExt::MatrixMatrix::add() sorted symbolic")));
+#endif
+  KokkosSparse::Experimental::spadd_symbolic
+    <KKH,
+    typename row_ptrs_array::const_type, typename col_inds_array::const_type,
+    typename row_ptrs_array::const_type, typename col_inds_array::const_type,
+    row_ptrs_array, col_inds_array>
+      (&handle, Arowptrs, Acolinds, Browptrs, Bcolinds, Crowptrs);
+  Cvals = values_array("C values", addHandle->get_max_result_nnz());
+  Ccolinds = col_inds_array("C colinds", addHandle->get_max_result_nnz());
+#ifdef HAVE_TPETRA_MMM_TIMINGS
+  MM = Teuchos::null;
+  MM = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("TpetraExt::MatrixMatrix::add() sorted kernel: sorted numeric")));
+#endif
+  KokkosSparse::Experimental::spadd_numeric(&handle,
+    Arowptrs, Acolinds, Avals, scalarA,
+    Browptrs, Bcolinds, Bvals, scalarB,
+    Crowptrs, Ccolinds, Cvals);
+}
+
+template<typename GO,
+         typename LocalIndicesType,
+         typename GlobalIndicesType,
+         typename ColMapType>
+struct ConvertColIndsFunctor
+{
+  ConvertColIndsFunctor (const GO minGlobal_,
+                         const LocalIndicesType& colindsOrig_,
+                         const GlobalIndicesType& colindsConverted_,
+                         const ColMapType& colmap_) :
+    minGlobal (minGlobal_),
+    colindsOrig (colindsOrig_),
+    colindsConverted (colindsConverted_),
+    colmap (colmap_)
+  {}
+  KOKKOS_INLINE_FUNCTION void
+  operator() (const size_t& i) const
+  {
+    colindsConverted[i] = colmap[colindsOrig[i]];
+  }
+  GO minGlobal;
+  LocalIndicesType colindsOrig;
+  GlobalIndicesType colindsConverted;
+  ColMapType colmap;
+};
+
+template<typename SC, typename LO, typename GO, typename NO>
+void MMdetails::AddKernels<SC, LO, GO, NO>::
+convertToGlobalAndAdd(
+  const typename MMdetails::AddKernels<SC, LO, GO, NO>::KCRS& A,
+  const typename MMdetails::AddKernels<SC, LO, GO, NO>::impl_scalar_type scalarA,
+  const typename MMdetails::AddKernels<SC, LO, GO, NO>::KCRS& B,
+  const typename MMdetails::AddKernels<SC, LO, GO, NO>::impl_scalar_type scalarB,
+  const typename MMdetails::AddKernels<SC, LO, GO, NO>::local_map_type& AcolMap,
+  const typename MMdetails::AddKernels<SC, LO, GO, NO>::local_map_type& BcolMap,
+  GO minGlobalCol,
+  GO /* numGlobalCols */,
+  typename MMdetails::AddKernels<SC, LO, GO, NO>::values_array& Cvals,
+  typename MMdetails::AddKernels<SC, LO, GO, NO>::row_ptrs_array& Crowptrs,
+  typename MMdetails::AddKernels<SC, LO, GO, NO>::global_col_inds_array& Ccolinds)
+{
+  using Teuchos::TimeMonitor;
+
+  const values_array& Avals = A.values;
```

Comment:
Be careful handling Views by reference, since somebody might insert code below that captures them in a device kernel.  Prefer handling them by value.